operators/pkg/controller/elasticsearch/elasticsearch_controller.go

```diff
@@ -283,3 +297,78 @@ func (r *ReconcileElasticsearch) finalizersFor(
+
+	// if the annotation does not exist, it might indicate it was reconciled by an older controller version that did not add the version annotation,
+	// in which case it is incompatible with the current controller, or it is a brand new resource that has not been reconciled by any controller yet
+	if !annExists {
+		exist, err := r.checkExistingResources(es)
+		if err != nil {
+			return false, err
+		}
+		if exist {
+			log.Info("Resource was previously reconciled by incompatible controller version and missing annotation, adding annotation", "controller_version", r.OperatorInfo.BuildInfo.Version, "namespace", es.Namespace, "es_name", es.Name)
+			err = annotation.UpdateControllerVersion(r.Client, es, "0.8.0-UNKNOWN")
+			return false, err
+		}
+		// no annotation exists and there are no existing resources, so this has not previously been reconciled
+		err = annotation.UpdateControllerVersion(r.Client, es, r.OperatorInfo.BuildInfo.Version)
+		return true, err
+	}
+
+	// if we have an annotation we need to check if it is compatible
+	currentVersion, err := semver.NewVersion(es.Annotations[annotation.ControllerVersionAnnotation])
+	if err != nil {
+		return false, errors.Wrap(err, "Error parsing current version on resource")
+	}
+	minVersion, err := semver.NewVersion("0.9.0-ALPHA")
+	if err != nil {
+		return false, errors.Wrap(err, "Error parsing minimum compatible version")
+	}
+	ctrlVersion, err := semver.NewVersion(r.OperatorInfo.BuildInfo.Version)
+	if err != nil {
+		return false, errors.Wrap(err, "Error parsing controller version")
+	}
+
+	// if the current version is gte the minimum version then they are compatible
+	if currentVersion.GreaterThanOrEqual(minVersion) {
+		log.V(1).Info("Current controller version on resource is compatible with running controller version", "controller_version", ctrlVersion.String(),
+			"resource_controller_version", currentVersion.String(), "namespace", es.Namespace, "es_name", es.Name)
+		return true, nil
+	}
+
+	log.Info("Resource was created with older version of operator, will not take action", "controller_version", ctrlVersion.String(),
+		"resource_controller_version", currentVersion.String(), "namespace", es.Namespace, "es_name", es.Name)
+	return false, nil
+}
+
+// checkExistingResources returns a bool indicating if there are existing resources created for a given resource
+func (r *ReconcileElasticsearch) checkExistingResources(es *elasticsearchv1alpha1.Elasticsearch) (bool, error) {
+	// if there's no controller version annotation on the ES instance, then we need to see maybe the CR has been reconciled by an older, incompatible controller version
+	selector := labels.NewSelector()
+	req, err := labels.NewRequirement(label.ClusterNameLabelName, selection.Equals, []string{es.Name})
+	if err != nil {
+		return false, err
+	}
+	selector.Add(*req)
+	opts := ctrlclient.ListOptions{
+		LabelSelector: selector,
+		Namespace:     es.Namespace,
+	}
+	pods := v1.PodList{}
```

Comment:
Does it make sense to select services here instead, which is common to all three APM, Kibana and ES

Reply:
Sure, that makes sense to me